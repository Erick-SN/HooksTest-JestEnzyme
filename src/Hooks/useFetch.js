import { useState, useEffect, useRef } from 'react';

export const useFetch = (url) => {
  const isMount = useRef(true);
  useEffect(() => {
    return () => {
      isMount.current = false;
    };
  }, []);
  const [state, setState] = useState({
    data: null,
    loading: true,
    error: null,
  });
  useEffect(() => {
    fetch(url)
      .then((res) => res.json())
      .then((data) => {
        if (isMount.current) {
          setState({
            loading: false,
            error: null,
            data,
          });
        }
      });
  }, [url]);
  return state;
};
