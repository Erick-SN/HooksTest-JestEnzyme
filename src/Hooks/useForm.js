import { useState } from 'react';

export const useForm = (initialState) => {
  const [value, setValue] = useState(initialState);
  const inputHandleChange = ({ target }) => {
    setValue({
      ...value,
      [target.name]: target.value,
    });
  };
  return [value, inputHandleChange];
};

export default useForm;
