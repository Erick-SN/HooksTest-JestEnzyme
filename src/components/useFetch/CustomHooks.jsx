import React from 'react';
import { useFetch } from '../../Hooks/useFetch';
import { useCounter } from '../../Hooks/useCounter';
import '../useEffect/effect.css';
const CustomHooks = () => {
  const { counter, increment } = useCounter(1);
  const { loading, data } = useFetch(
    `https://www.breakingbadapi.com/api/quotes/${counter}`,
  );
  const { author, quote } = !!data && data[0];
  return (
    <>
      <h1>BreakingBad Quotes</h1>
      <hr />
      {loading ? (
        <div className='alert alert-info text-center'>Loading...</div>
      ) : (
        <blockquote className='blockquote text-right'>
          <p className='mb-0'>{quote}</p>
          <footer className='blockquote-footer'>{author}</footer>
        </blockquote>
      )}
      <button onClick={() => increment()} className='btn btn-primary'>
        Next
      </button>
    </>
  );
};

export default CustomHooks;
