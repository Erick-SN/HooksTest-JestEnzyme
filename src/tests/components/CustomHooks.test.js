import React from 'react';
import { shallow } from 'enzyme';
import { useFetch } from '../../Hooks/useFetch';
import { useCounter } from '../../Hooks/useCounter';
import CustomHooks from '../../components/useFetch/CustomHooks';
jest.mock('../../Hooks/useFetch');
jest.mock('../../Hooks/useCounter');

describe('Custom Hooks test', () => {
  useCounter.mockReturnValue({
    counter: 1,
    increment: () => {},
  });

  test('should render component correctly', () => {
    useFetch.mockReturnValue({
      data: null,
      loading: true,
      error: null,
    });
    const wrapper = shallow(<CustomHooks />);
    expect(wrapper).toMatchSnapshot();
  });

  test('should show information', () => {
    useFetch.mockReturnValue({
      data: [
        {
          author: 'Erick',
          quote: 'Hello world',
        },
      ],
      loading: false,
      error: null,
    });
    const wrapper = shallow(<CustomHooks />);
    expect(wrapper.find('.alert').exists()).toBe(false);
    expect(wrapper.find('.mb-0').text().trim()).toBe('Hello world');
  });
});
