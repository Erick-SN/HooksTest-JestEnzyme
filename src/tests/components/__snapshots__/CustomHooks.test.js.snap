// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Custom Hooks test should render component correctly 1`] = `
<Fragment>
  <h1>
    BreakingBad Quotes
  </h1>
  <hr />
  <div
    className="alert alert-info text-center"
  >
    Loading...
  </div>
  <button
    className="btn btn-primary"
    onClick={[Function]}
  >
    Next
  </button>
</Fragment>
`;
